compiler/rustc_hir_typeck/src/demand.rs

```diff
@@ -286,6 +286,19 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                     err.downgrade_to_delayed_bug();
                 }
             }
+            (
+                Some(hir::Node::Expr(hir::Expr {
+                    kind: hir::ExprKind::Binary(_, lhs, rhs), ..
+                })),
+                Some(TypeError::Sorts(ExpectedFound { expected, .. })),
+            ) if rhs.hir_id == expr.hir_id
+                && self.typeck_results.borrow().expr_ty_adjusted(lhs) == expected =>
```

Comment:
this shouldn't happen, but you should probably use `expr_ty_adjusted_opt` here just to avoid ICEs